### ISSUE TITLE
Performance - Overhead of using Realm objects

### ISSUE THREAD

**Issue by l7x** [realm/realm-js#2782]
Currently, using Realm in a scenario which require a lot of calls through the C++/JS boundary (i.e. in a tight loop), can totally kill the performance of an app.

We did some tests and for example, just accessing something like

`results[0].myprop`

was >200x slower with Realm objects than normal JS array of objects (which you'd get after running a query on other databases).

Some possible solutions:

- [Graal.js](https://github.com/realm/realm-js/issues/2756) in theory could help with that, but it seems that node extensions are not optimized at this point

- Automatic caching of retrieved data (datomic does something similar)

- Using something like JSON.parse(JSON.stringify(results)) instead of realm collections and objects - slow and ugly, last-resort hack

**Comment by blagoev** [comment]
Accessing a realm object's property is more involved than accessing a POJO property. Realm Object properties should always reflect the real data in the database. If you need an array of objects to query very heavily then you are on the right track to clone the realm objects into pojo objects and have them in an array. But normally you wouldn't need to do that.

Crossing the c++/js barrier is done by the JS virtual machine which behaves very differently when accessing a plan old js object with no native counterpart. So comparing these is not correct.

**Comment by bmunkholm** [comment]
@i7x could you describe what exactly you are doing and why that causes a performance concern?

**Comment by l7x** [comment]
@bmunkholm The simplest example is basically a tight loop where you're accessing realm collections/object props (i.e. iterating over many rows and calculating some data based on properties).

@blagoev I understand the difference between these two things - I opened this issue to see if there's a way to minimize the performance impact of that overhead, so I don't agree that comparing these two things in this way is necessarily "incorrect".

Yes, cloning collections is first what comes to mind as potential solution, but with realm objects that can get very messy, very fast - realm objects are (recursively) nested directly through references, they are not that easily converted to POJOs (like i.e. SQLite rows).
Maybe some efficient methods on collections can make this a bit less painful, but I'm wondering if there are better solutions - i.e. IIRC Datomic is doing automatic caching of retrieved data, maybe Realm.js could (optionally) do something similar on JS side.

**Comment by bmunkholm** [comment]
@l7x the reason I ask for a more specific example is because iterating over properties and e.g. aggregating values is super fast in the core database. So if you e.g. want to sum, avg or count properties that's easily handled by the core db and you should use queries for that. But there surely are many operations not supported by the core database. It would also be good to understand the size of collections that cause problems. 
Could you show an example of operations you perform, the size of the collections and the time it takes?
It may be that we should rephrase the title of this issue something like "Iterating over large lists of properties is slow" or something like that? It's not clear that it's the C++/JS interface versus what's done on the C++ side that is slow?

**Comment by l7x** [comment]
It's a proprietary app and we're not allowed to share the code, but it's very easy to simulate like I mentioned before.
Yes, doing trivial avg/sum/count on collections is super fast (there's just one call to C++), but there'll always be cases where you'll need to do the iteration and calculation yourself.

It's not just the size of the collection - it's the collection-size * number-of-native-calls (so, i.e. if you have a lot of property access per loop iteration, it can be very noticeable even in smaller collections).

Regarding the changing of the title, I don't know - it's not really a secret that crossing the C++/JS barrier has some overhead. It's not just realm, I've seen that before - C++ is fast, JS is fast, but when you're start making a lot of calls between them, things become slow.
I haven't dug deep enough to say for sure, but it seems that Realm's C++ side is very well optimized, or the difference would be higher.

**Comment by l7x** [comment]
Just to clarify things a bit more - this is a tricky issue and I don't think there is a good quick solution. I opened this issue to see if there's something that can be done in the long term (in the meantime, we can use different hacks to minimize the impact). Because otherwise we have to rewrite the whole app to use SQLite (which works better in such scenarios) - something I don't look forward to.

**Comment by bmunkholm** [comment]
Yeah I get that you can't share the exact code. But perhaps you can describe at the abstract level what you need to do. The reason I ask is that you are the first (I can recall) that have mentioned this as an actual practical issue. So in case you need other aggregation operations, it's likely easier to add those than optimize the whole property assessor path (although we should surely look to see if there are easy wins). What to do also depends on if we deem it needed to cut the time by a factor of 2, 10 or 100 :-) to make it sufficiently performant for what you need.
Depending on what you do/need it could also be that the solution is to make it easy(er) for you to actually pull out the properties to native JS objects.
So I would still appreciate if you could let us know what you need to do in that loop and what performance would be sufficient for your use case.

**Comment by l7x** [comment]
I understand what you're saying and in some simpler case, it would make a lot of sense to ask you for some fast aggregate function, but this is way too complex - scores are calculated based on a lot of properties and also on relationships between different items. To make things worse, those rules are also subject to change.
It's definitely impossible to abstract that into some generic aggregate functions.

But now that you mentioned it, we have a different cases (unrelated to this), where "aggregate" functions on collections could really help. Something like results.pluck("id") which would return something like ["key1", "key2", ...] and another function to return something like {"key1": obj, "key2": obj, ...}. Such native functions should be much faster than we can currently get when working with realm objects in JS.

**Comment by blagoev** [comment]
If the task you have can't be accomplished by the database layer by a query with an aggregate function then you need to fetch the object from the database and that will cross the boundary regardless what database is used. 
I still want to emphasize that the expectation this to be as fast as accessing a cached JS array are not correct. 
Fetching and caching the results set makes only sense if you are going multiple passes over the collection. If you need to do this only once then looping over the results set and accessing every object will be as fast as caching the object.

In other words on your question: Is there a way to minimize the number of c++/js transitions?
The answer is try using the aggregate query functions as @bmunkholm have said. If this is not possible then the other option is to restructure the data to support the specific scenario and store it in a form so it's easy to query with an aggregate function. It all depends on the task at hand. 

Hope this helps.

**Comment by l7x** [comment]
@blagoev The generalizations you wrote are not correct. With SQLite (and pretty much any other database), you're generally crossing the boundary once to fetch the rows.
With Realm, as you know, you're crossing the boundary every time you're accessing the rows in the collection and for EVERY property access on the row. That's very different. If you heavily use those objects, there'll be a lot of overhead.

Cloning those objects is also very messy with Realm, as I wrote in my previous comment.


I think we'll rewrite this project to use SQLite. We may still use Realm.js in some lighter scenarios. IMHO, in current form it's not a good fit for something that requires a heavy use of the domain objects.

Thank you for your answers.

**Comment by blagoev** [comment]
@l7x Could you share some more details on how you are seeing 200x performance difference when accessing Realm objects versus JS objects. It's not obvious from the example you provided

Accessing `results[0].myprop` property should not hit such big performance degradation. On what platform this is happening and what are the details below:

- Realm JS SDK Version: ?
- Node or React Native: ?
- Client OS & Version: ?
- Which debugger for React Native: ?/None

**Comment by l7x** [comment]
Sure, that was the whole point of opening this issue, but there was a lot of "losing sight of the forest for the trees" in this thread. We tested it on our codebase, but I'll create some minimal general test case.

**Comment by l7x** [comment]
I created a simple test case based on your "Getting started" tutorial:

Realm JS SDK Version: 5.0.2
Node or React Native: Node 12 & 10
Client OS & Version: Debian Linux 9


Note, benchmark.js can use a lot of memory, so you'll likely need to start it with something like this:
`node --max-old-space-size=8192 test.js`

test.js
```
const Realm = require('realm');
var Benchmark = require('benchmark');
var suite = new Benchmark.Suite;

const CarSchema = {
  name: 'Car',
  properties: {
    make:  'string',
    model: 'string',
    miles: {type: 'int', default: 0},
  }
};

Realm.open({schema: [CarSchema]})
  .then(realm => {
    realm.write(() => {
      const myCar = realm.create('Car', {
        make: 'Honda',
        model: 'Civic',
        miles: 2000,
      });
    });

    const cars = realm.objects('Car')
    const jcars = [{
        make: 'Honda',
        model: 'Civic',
        miles: 2000,
      }]
      

	let rres = ''
	let jres = ''
	suite.add('realm', function() {
		rres += cars[0].make + cars[0].model
	})
	suite.add('js', function() {
		jres += jcars[0].make + jcars[0].model
	})	
	.on('cycle', function(event) {
	  console.log(String(event.target));
	})
	.run();

    realm.close();
    
  })
  .catch(error => {
    console.log(error);
  });

```

Result on my VM:
```
realm x 49,765 ops/sec �1.82% (91 runs sampled)
js x 4,115,300 ops/sec �29.96% (56 runs sampled)
```

Like I sad before, I understand that this type of test is not really "fair" towards realm (accessing props on realm object is much more involved and it can't be easily optimized by VM), but at the end of the day, both with SQLite and Realm (or anything else), you always end up accessing those properties (sometimes heavily), so with that in mind, I don't think it's necessarily "incorrect".

**Comment by cristianoccazinsp** [comment]
Could it be that you're calling `cars[0]` every time? What about assigning it first? `let c= cars[0]` and then do your operations with `c`.

**Comment by l7x** [comment]
@cristianoccazinsp Sure, it's easy to modify the test to hit only props (our original tests were pretty much doing that).
This test is hitting the previously mentioned `results[0].myprop` access.
I think a lot of people don't understand hidden costs of abstractions they use with realm. There's a lot of benchmarks out there showing how much it's faster than SQLite with very light use of collections and objects, but none of them is showing what happens with a heavy use.

**Comment by cristianoccazinsp** [comment]
Just to be clear, hitting only props doesn't have such issue? For me it would be pretty obvious that accessing `results[0]` will incurr in the query re-executing every time, isn't that what most ORMs do? Queries are lazy, if you want an efficient access, just `map` over the result set and build a POJO of just the properties you need.

This is what I usually do, although with Realm 5.x there's probably a better way to do this with the usage of `keys()` and `entries()`.
```
let res = realm.objects('schema').filtered(...).sorted(...).map(obj => toPlain(obj, schema.properties));

// note: we could only map a subset of the schema properties if we are only
// interested in some
const toPlain = function(obj, schema){
  let res = {};
  for(let k of Object.keys(schema)){
    res[k] = obj[k]
  }
  return res;
}

**Comment by l7x** [comment]
No, you misunderstood me. Hitting props also has this issue. Just try to replace cars[0] with a local variable and try to access props many times in a loop, this is my result:
```
realm x 31,250 ops/sec �0.49% (95 runs sampled)
js x 78,704,423 ops/sec �0.81% (85 runs sampled)
```

**Comment by cristianoccazinsp** [comment]
Ok, then that's definitely different to what your typical ORM would provide. ORMs usually cache props access in case you access it multiple times. Does this mean Realm hit the database each time the prop is accessed? If that's documented, then I think it's an acceptable caveat.

Either way, if you just "convert" your objects to a list/pojo then you shouldn't have this issue?

**Comment by l7x** [comment]
No, ORMs are typically "caching" whole resultsets.
Also, realm isn't really hitting the database for prop access, they are cached, but that require crossing the JS/C++ barrier.
Also, you mentioned earlier that accessing results[0] will re-execute the query - that's also not strictly true, queries are also cached, but that access also requires crossing the JS/C++ barrier.

**Comment by l7x** [comment]
Converting everything to POJOs is one solution, but I mentioned in my previous comments why it usually becomes messy with realm.

**Comment by cristianoccazinsp** [comment]
That's actually quite good to know.  I haven't really gone through the docs to chase these "caveats" since I've never experienced such bottlenecks, but it's good to know.

"No, ORMs are typically fetching whole resultsets."  You are right with this. I'm comparing with django where you can optionally "lazy load" properties so that each time you access them they are retrieved from the database.

I can't really think of another alternative than converting to plain JS objects. I recall having issues of all kinds with realm objects when they were used in a "non realm" context, so from there I've always been using plain JS objects.

**Comment by l7x** [comment]
Are you sure django is lazy loading properties? It's been long time since I used it, but I can't imagine that being very efficient even with something like SQLite, not to mention mysql/postgresql.

**Comment by cristianoccazinsp** [comment]
Not by default, but you can tell it which properties to "defer". That's especially useful when you just want to use 1 or 2 props from the database so you don't query the whole row into memory. 

Now I wonder, does realm load the whole "database row" into memory but lazily gives it to JS through the C++/JS bridge, or does it lazy load each property from the database?

Lastly, so we are confident to say that realm does not cache any prop on the JS side, and it always hits the C++ side when accessing a property? I can see this being a huge caveat when coming from typical ORMs.

**Comment by l7x** [comment]
Realm is more column oriented and using mmap caching. You can find more info on their website.

I do have some ideas about some helper methods which may make converting to POJOs less messy, but I'm in a hurry now, I'll write them here later.

**Comment by l7x** [comment]
So, let's say that we have a typical one-to-many relation. In SQLite, you'll have something like this:
```
users:
[{
	id: 1,
}]
items:
[{
	id: 1,
	user_id: 1
}]
```

It's very easy to fetch items POJOs and when/if you need related user, you can also easily fetch a user POJO.


Now with realm, you have a tight coupling with object references and you can't easily fetch just items POJOs:
```
items:
[{
	id: 1,
	user: {
		id: 1,
		items: *recursion*
	}
}]
```

So we need a some kind of native (efficent) `results.fetch()` (or whatever you want to call it) method, which will convert a realm collection to js array of POJOs with foregin keys instead of object references. It's tricky because because realm doesn't really have foreign keys, so there would need to be some kind of naming convention about returned "foreign keys" on POJOs (or have them same names as their object counterparts, but that may be confusing).
I think that's pretty important feature, because AFAIK currently there's no good way to even export your collection to JSON (and back).

SQLite (and other relational dbs) doesn't have this problem because its generally more loosely coupled (it would be nice to be able to do a dynamic join also in realm in the future).

**Comment by l7x** [comment]
I changed the title of the issue so it will be easier for users to understand what it's about by just looking at the title.

**Comment by blagoev** [comment]
@l7x  Thanks for the provided sample. 

At first glance while the code for both is exactly the same, there are the following differences:
the `cars[0]` call is one ffi call which retrieves data an creates an object from the database
if you do this multiple times it will result in multiple different objects so 
```
var equals = cars[0]  === cars[0] //is false
```
this is a design decision in Realm to always provide up-to-date "live" objects from the database

Also 
```
var mycar = cars[0]; //one ffi transition
mycar.make` //second ffi transition
```
so the code 
`cars[0].make + cars[0].model` is actually 4 ffi transitions in total. 

Also from the example above 
`jcars[0].make` is accessing an in memory property which v8 can optimize very good. 
`cars[0].make` is invoking a property accessor on the prototype of the cars[0] object. And this accessor is a doing the ffi transition to retrieve the data. 

One suggestion is to use a temporary variable when accessing an object from the array
```
var car = cars[0];
var x = car.name + car.make;
```

Even then memory access is going to be fast, in fact much faster since of the v8 hot path compilation and local caches etc.

Comparing Realm to other databases is much fair. But then we should remember the design differences. 
Indeed Realm will make a second access for the `make` value but that's in the heart of Realm to provide always up-to-date property values, live objects and change notifications.
`car.name + car.make` is 2 ffi calls

Cloning the full object with all properties in memory on retrieval could be slower if the access is on single property since the whole object is serialized into memory, and it's consuming more memory to cache the object. 
CPU vs Memory is always a trade of when talking performance. Being a mobile database Realm is trying to hit a good balance between those two.

Hope the suggestion above is helpful.

**Comment by l7x** [comment]
Yeah, I explained that in my reply to @cristianoccazinsp. This test is hitting the previously mentioned results[0].myprop access (which also reveals another issue which we noticed and you mentioned - if you access results[0] multiple times, why not return the same object (it's a "live" object anyway), instead of creating a new one?).
Modifying the test to only test hitting a lot of properties in a loop is trivial - just copy&paste a property access many times in the loop.

https://github.com/realm/realm-js/issues/2782#issuecomment-608443330
https://github.com/realm/realm-js/issues/2782#issuecomment-608453197

**Comment by l7x** [comment]
Regarding v8 optimizations - yes, I mentioned that in my comment bellow the test, but in a way it also proves my point - there's a lot of optimizations that simply can't be done on realm objects due to the way things currently work.

Anyway, I think it's worth at least considering some of the options mentioned in this thread. Due to the way things currently work, most developers simply don't understand the abstractions they're using and costs they are paying basically on every access of those very fundamental abstractions.

**Comment by RealmBot** [comment]
➤ Lyubomir Blagoev commented:

closing this issue as no further actions are pending on it.
